content/blog/2025-04-24-super-optimization/index.md

```diff
@@ -0,0 +1,104 @@
+
+**STOKE: Stochastic Superoptimization (2013)**
+
+
+The STOKE superoptimizer formulates the superoptimization task as a stochastic search problem and search with MCMC. The cost function is defined as a combination of correctness and perf improvement. This approach trades off between completeness and the size of search space. Some human annotations are required to help generate random program input used to approximate the cost function value. The level of human intervention is considered to be light work compared with the other two approaches we mentioned and expert knowledge is not required. This approach is able to handle CISC ISA (x86_64) with over 400 opcodes.
+
+
+**Souper: A Synthesizing Superoptimizer (2018)**
+
+Souper takes a different approach to the other superoptimizers discussed, in that instead of outputting assembly, its output is IR: it both takes in, and produces as output, LLVM IR. Its algorithm is CEGIS: enumeration and testing, like Massalin’s superoptimizer, but in a loop with a verification tool that can add tests. It’s designed to be used mainly by compiler engineers, to help improve compiler optimizations: both LLVM and MSVC include optimizations inspired by suggestions from the output of Souper.
+
+There is a body of other superoptimization work focusing on more domain specific settings ([tensor computation](https://arxiv.org/abs/2101.01332) optimization in a deep learning setting) and more restricted search spaces ([SIMD instructions](https://arxiv.org/abs/2306.00229)).
+
+## Works cited
+
+H. Massalin. Superoptimizer - a look at the smallest program. In ASPLOS, pages 122–126, 1987.
+
+Solar-Lezama, Armando, et al. "Combinatorial sketching for finite programs." Proceedings of the 12th international conference on Architectural support for programming languages and operating systems. 2006.
+
+Gulwani, Sumit, et al. "Synthesis of loop-free programs." ACM SIGPLAN Notices 46.6 (2011): 62-73.
+
+Bansal, Sorav, and Alex Aiken. "Automatic generation of peephole superoptimizers." ACM SIGARCH Computer Architecture News 34.5 (2006): 394-403.
+
+Sasnauskas et al. “Souper: A Synthesizing Superoptimizer” (2018). https://arxiv.org/abs/1711.04422.
+Schkufza, Eric, Rahul Sharma, and Alex Aiken. "Stochastic superoptimization." ACM SIGARCH Computer Architecture News 41.1 (2013): 305-316.
+
+Yang, Phothilimtha et al. “Equality Saturation for Tensor Graph Superoptimization” (2021). https://arxiv.org/abs/2101.01332
+
+Liu, Mada and Regehr. “Minotaur: A SIMD-Oriented Synthesizing Superoptimizer” (2024). https://arxiv.org/abs/2306.00229
+
+
```

Comment:
Because this is the web, I think you can cut out this list and just use hyperlinks instead!